sympy/functions/elementary/tests/test_trigonometric.py

```diff
@@ -341,6 +343,8 @@ def test_cos():
 
     assert cos(x*I) == cosh(x)
     assert cos(k*pi*I) == cosh(k*pi)
+    assert cos(k*pi + (m - 1)*pi/2 + 1) == -sin(1)*sin(pi*(m/2 + k - S(1)/2)) + \
+        cos(1)*cos(pi*(m/2 + k - S(1)/2))
```

Comment:
I'm not sure this is an improvement. We have:
```julia
In [20]: cos(1+m*pi)                                                                                                   
Out[20]: 
    m       
(-1) ⋅cos(1)
```
Okay that seems reasonable. I do wonder if that should always evaluate though.

Then we have
```julia
In [21]: cos(1+m*pi/2)                                                                                                 
Out[21]: 
            ⎛π⋅m⎞             ⎛π⋅m⎞
- sin(1)⋅sin⎜───⎟ + cos(1)⋅cos⎜───⎟
            ⎝ 2 ⎠             ⎝ 2 ⎠
```
In that case I think I prefer the behaviour on master:
```julia
In [1]: cos(1+m*pi/2)                                                                                                  
Out[1]: 
   ⎛π⋅m    ⎞
cos⎜─── + 1⎟
   ⎝ 2     ⎠
```

Reply:
Both things can not be made to work simultaneously I guess.

Reply:
On master:
```
In [3]:  from sympy.functions.elementary.trigonometric import _peeloff_pi as peel

In [4]: m = symbols('m', integer=True)

In [5]: peel(pi*m + 1)
Out[5]: (pi*m + 1, 0)

In [6]: peel(1+m*pi/2)
Out[6]:
 pi*m
(---- + 1, 0)
  2
```
On my branch:
```
In [1]: from sympy.functions.elementary.trigonometric import _peeloff_pi as peel

In [2]: m = symbols('m', integer=True)

In [3]: peel(pi*m + 1)
Out[3]: (1, pi*m)

In [4]: peel(1+m*pi/2)
Out[4]:
    pi*m
(1, ----)
     2
```

The behaviour on my branch seems better from the point of view of the description of `_peeloff_pi` as mentioned in the docstrings.

```
    Split ARG into two parts, a "rest" and a multiple of pi/2.
    This assumes ARG to be an Add.
    The multiple of pi returned in the second position is always a Rational.
```

Reply:
> I'm not sure this is an improvement.

In that case, I guess that this should not evaluate automatically.
So, as mentioned on the issue page, the expected simplification happens after applying `trigsimp()` and thus there will be nothing left to resolve in this issue.

On master:
```
In [1]: n = symbols('n', integer=True)

In [2]: sin(2*n*pi + 4)
Out[2]: sin(2*pi*n + 4)

In [3]: _.trigsimp()
Out[3]: sin(4)
```

Reply:
There is always a tension in sympy between automatic evaluation on the one hand and preserving the inputs on the other. It would be better in general if we could preserve the inputs a lot more but that isn't how things work currently so there is often an expectation that things should "evaluate" more if possible. It's hard to say what the balance should be but we're somewhere in the middle right now which means we have the (dis)advantages of both approaches.

Reply:
```
In [8]: m = symbols('m', rational=True)

In [9]: peel(1+m*pi/2)
Out[9]:
    /m        \      pi*m      /m        \
(pi*|- mod 1/2| + 1, ---- - pi*|- mod 1/2|)
    \2        /       2        \2        /

```

Reply:
I think we should restrict ourselves to `integers`, otherwise there can always be some cases which might cause a `Recursion Error`, possibly because of the result of `mod` remaining unevaluated.

Reply:
Secondly, I feel that the problem on the issue page was more highlighting this behaviour:
```
In [1]: n = symbols('n', integer=True)

In [2]: sin(2*n*pi + 4)
Out[2]: sin(2*pi*n + 4)

In [3]: _.simplify()
Out[3]: sin(2*pi*n + 4)

In [4]: _.trigsimp()
Out[4]: sin(4)
```

Reply:
I guess, we should rather not focus on making this evaluation happen automatically.